### ISSUE TITLE
read charge and vdw paramaters

### ISSUE THREAD

**Issue by avdisalem** [openmm/openmm#3019]
Hi,

Is it possible to read the charge, sigma and epsilon values of atoms of a molecule (e.g. benzene) from the str file file CHARMM Drude Force Field? How can I do that?

**Comment by peastman** [comment]
I think this is a CHARMM question, not an OpenMM question?  The CHARMM file formats are documented in https://www.charmm.org/charmm/documentation/by-version/c42b1/params/doc/io/.

**Comment by avdisalem** [comment]
I think it's an OpenMM question because I am using OpenMM to run the simulations. I would read the charge, sigma and epsilon from the str file of CHARMM and use them in a modified van der waals potential using OpenMM. I think I should use `getParticleParameters` and `addPerParticleParameter` commands but I don't know how to use them.

**Comment by peastman** [comment]
Could you give more information about what you're doing?  Are you using CharmmPsfFile to read a PSF file, then calling `createSystem()` on it to a create an OpenMM System?  If so, you can query information from the System, but that will be after it's already created.  Since you want to use a modified potential, it sounds like you may be reading the files yourself and writing your own code to build the System?  But I can't tell if that's actually what you're doing.

**Comment by avdisalem** [comment]
Yes, I am reading the psf and pdb files, then calling `createSystem` as follows:
```
params = app.CharmmParameterSet('toppar_drude_master_protein_2019g.str')
system = psf.createSystem(params, nonbondedMethod=ff.PME, nonbondedCutoff=1.2*unit.nanometer, switchDistance=1.0*unit.nanometer, ewaldErrorTolerance = 0.0001, constraints=ff.HBonds)
```
If I can read the the charges, sigmas and epsilons from the `toppar_drude_master_protein_2019g.str`, then I would use them in a modified potential. Suppose that system consists of one molecule benzene and 512 water molecules.

**Comment by ahmedo047** [comment]
Hi,

I have a similar problem. The script I use works properly when I read the `amber99sbildn.xml` force field file to get the charge, sigma and epsilon paramaters of methanol atoms (see the below code snippet). It is interesting that the modified script gives the value error (`ValueError: AmoebaUreyBradleyGenerator: error getting types: CD2R5B ND2R5D HDP1A`) while using the CHARMM Drude Force Field (`charmm_polar_2019.xml`). Any ideas on how to fix this error? Note that I use the below code snippet that works with `amber99sbildn.xml` for `charmm_polar_2019.xml` as well.
```
param_charge_list, param_sigma_list, param_epsilon_list = [], [], []
for i in range(none_bonded_force.getNumParticles()):
    charge, sigma, epsilon = none_bonded_force.getParticleParameters(i)
    param_charge_list.append(charge)
    param_sigma_list.append(sigma)
    param_epsilon_list.append(epsilon)
```

**Comment by peastman** [comment]
The information you want is stored in a few different places, split between the CharmmPsfFile and CharmmParameterSet.

`psf.atom_list` is a list of [Atom](https://github.com/openmm/openmm/blob/6d4abc54201b06581337fbca33349fae6279843a/wrappers/python/openmm/app/internal/charmm/topologyobjects.py#L239) objects for the atoms.  The charges are stored in those objects:

```python
>>> psf.atom_list[0].charge
-0.3
```

The LJ parameters are stored in the CharmmParameterSet.  They're indexed by atom type, which you get from the `Atom`.

```python
>>> type = params.atom_types_int[psf.atom_list[0].attype]
>>> type.epsilon
-0.2
>>> type.rmin
1.85
```

Things to note about these values:

- Epsilon is in kcal/mol.
- Rmin is in Angstroms.
- NonbondedForce expects sigma rather than rmin.

You can do all the proper conversions like this.

```python
sigmaScale = 2**(-1/6) * 2
nonbonded.addParticle(charge, sigmaScale*rmin*angstroms, epsilon*kilocalories_per_mole)
```

- In general, the interaction strength may be reduced for 1-4 interactions.  But the CHARMM 2019 Drude force field sets the scales to 1, so that isn't an issue for it.
- These are the standard values.  CHARMM supports NBFIX terms, which cause different LJ parameters to be used between specific pairs of atom types.  CHARMM 2019 Drude does make use of this.  This information gets stored in the [`nbfix` field](https://github.com/openmm/openmm/blob/6d4abc54201b06581337fbca33349fae6279843a/wrappers/python/openmm/app/internal/charmm/topologyobjects.py#L99-L101) of the `AtomType`.

**Comment by jchodera** [comment]
@avdisalem : Did we address your question in sufficient detail? If so, can we close this issue?

**Comment by avdisalem** [comment]
**@jchodera** : Thanks for replies. You can close the issue of course.  But unfortunately I still couldn't read the charge, sigma and epsilon from the str file of CHARMM or  `charmm_polar_2019.xml`  and use those parameters in a modified van der Waals potential using OpenMM. It would be really good if you add such an example code to the example directory of OpenMM `https://github.com/openmm/openmm/tree/master/examples`.

**Comment by jchodera** [comment]
@avdisalem : The `str` files just contain parameters. Do you have a way of generating structures (e.g. PDB files with the appropriate `CONECT` records) for these molecules we could use to first load the systems and apply parameters within OpenMM? If you can provide an example molecule structure file, we can illustrate loading it, applying parameters, and modifying `charge`, `sigma`, `epsilon` in a simple example.

**Comment by avdisalem** [comment]
**@jchodera** : I attached an example molecule structure file (methanol in water). 
[solvated_meoh.zip](https://github.com/openmm/openmm/files/6574502/solvated_meoh.zip)
